pkg/operator/minio/controller.go

```diff
@@ -0,0 +1,304 @@
+					},
+					Spec: v1.PersistentVolumeClaimSpec{
+						AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+						Resources: v1.ResourceRequirements{
+							Requests: v1.ResourceList{
+								v1.ResourceStorage: resource.MustParse(minioStorageGBString),
+							},
+						},
+					},
+				},
+			},
+			ServiceName: name,
+			// TODO: liveness probe
+		},
+	}
+
+	return appsClient.StatefulSets(namespace).Create(&ss)
+}
+
+func (c *MinioController) makeMinioService(name, namespace string, spec miniov1alpha1.ObjectStoreSpec) (*v1.Service, error) {
+	coreV1Client := c.context.Clientset.CoreV1()
+
+	svc, err := coreV1Client.Services(namespace).Create(&v1.Service{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    map[string]string{"app": minioLabel},
+		},
+		Spec: v1.ServiceSpec{
+			Type:     v1.ServiceTypeNodePort,
+			Selector: map[string]string{"app": minioLabel},
+			Ports:    []v1.ServicePort{{Port: spec.Port}},
+		},
+	})
+
+	return svc, err
+}
+
+func (c *MinioController) onAdd(obj interface{}) {
+	objectstore := obj.(*miniov1alpha1.ObjectStore).DeepCopy()
+
+	// Create the headless service.
+	logger.Infof("Creating Minio headless service %s.", objectstore.Name)
+	_, err := c.makeMinioHeadlessService(objectstore.Name, objectstore.Namespace, objectstore.Spec)
+	if err != nil {
+		logger.Errorf("failed to create minio headless service: %v", err)
+		return
+	}
+	logger.Infof("Finished creating Minio headless service %s.", objectstore.Name)
+
+	// Create the stateful set.
+	logger.Infof("Creating Minio stateful set %s.", objectstore.Name)
+	_, err = c.makeMinioStatefulSet(objectstore.Name, objectstore.Namespace, objectstore.Spec)
+	if err != nil {
+		logger.Errorf("failed to create minio stateful set: %v", err)
+		return
+	}
+	logger.Infof("Finished creating Minio stateful set %s.", objectstore.Name)
+
+	// Create the nodeport service.
+	svcName := objectstore.Name + "-service"
+	logger.Infof("Creating Minio service %s.", svcName)
+	_, err = c.makeMinioService(svcName, objectstore.Namespace, objectstore.Spec)
+	if err != nil {
+		logger.Errorf("failed to create minio nodeport service: %v", err)
+		return
+	}
+	logger.Infof("Finished creating Minio service %s.", svcName)
+}
+
+func (c *MinioController) onUpdate(oldObj, newObj interface{}) {
+	oldStore := oldObj.(*miniov1alpha1.ObjectStore).DeepCopy()
+	newStore := newObj.(*miniov1alpha1.ObjectStore).DeepCopy()
+
+	_ = oldStore
+	_ = newStore
+	logger.Infof("Called onUpdate.")
+}
+
+func (c *MinioController) onDelete(obj interface{}) {
+	objectstore := obj.(*miniov1alpha1.ObjectStore).DeepCopy()
+	var delOpts meta_v1.DeleteOptions
+	propPolicy := meta_v1.DeletePropagationForeground
+	delOpts.PropagationPolicy = &propPolicy
+	//listOpts := meta_v1.ListOptions{LabelSelector: "app=" + minioLabel}
+	coreV1Client := c.context.Clientset.CoreV1()
+
+	// Delete nodeport service.
+	svcName := objectstore.Name + "-service"
```

Comment:
Instead of deleting the resources explicitly, you can set the `ownerReferences` field when you create resource. Then when the crd is deleted, all its resources will automatically be deleted